File: libwebrtc-sys/src/peer_connection.cc
#include "libwebrtc-sys/include/peer_connection.h"
#include "libwebrtc-sys/include/peer_connection_observer.h"
#include "libwebrtc-sys/include/peer_connection_stats_callback.h"
#include "libwebrtc-sys/src/data_channel.rs.h"
#include "libwebrtc-sys/src/peer_connection.rs.h"
#include "libwebrtc-sys/src/shared_bridge.rs.h"
#include "rust/cxx.h"
#include <iostream>
#include <vector>

void ArcasPeerConnection::create_offer(
    rust::Box<ArcasRustCreateSessionDescriptionObserver> observer) const
{
    webrtc::PeerConnectionInterface::RTCOfferAnswerOptions options;
    auto ref_counted =
        rtc::make_ref_counted<ArcasCreateSessionDescriptionObserver>(std::move(observer));
    api->CreateOffer(ref_counted, options);
}

void ArcasPeerConnection::create_answer(
    rust::Box<ArcasRustCreateSessionDescriptionObserver> observer) const
{
    // TODO: This probably has to live longer elsewhere.
    webrtc::PeerConnectionInterface::RTCOfferAnswerOptions options;
    auto ref_counted =
        rtc::make_ref_counted<ArcasCreateSessionDescriptionObserver>(std::move(observer));
    api->CreateAnswer(ref_counted, options);
}

void ArcasPeerConnection::set_local_description(
    rust::Box<ArcasRustSetSessionDescriptionObserver> observer,
    std::unique_ptr<ArcasSessionDescription> sdp) const
{
    auto ref_counted = rtc::make_ref_counted<ArcasSetDescriptionObserver>(std::move(observer));
    api->SetLocalDescription(std::move(sdp->clone_sdp()), ref_counted);
}

void ArcasPeerConnection::set_remote_description(
    rust::Box<ArcasRustSetSessionDescriptionObserver> observer,
    std::unique_ptr<ArcasSessionDescription> sdp) const
{
    auto ref_counted = rtc::make_ref_counted<ArcasSetDescriptionObserver>(std::move(observer));
    api->SetRemoteDescription(std::move(sdp->clone_sdp()), ref_counted);
}

std::unique_ptr<ArcasRTPVideoTransceiver> ArcasPeerConnection::add_video_transceiver() const
{
    auto result = api->AddTransceiver(cricket::MEDIA_TYPE_VIDEO);

    if (result.ok())
    {
        return std::make_unique<ArcasRTPVideoTransceiver>(*api, result.MoveValue());
    }

    // TODO: Handle error cases.
    return nullptr;
}

std::unique_ptr<ArcasRTPVideoTransceiver>
ArcasPeerConnection::add_video_transceiver_with_track(std::unique_ptr<ArcasVideoTrack> track,
                                                      ArcasTransceiverInit init) const
{
    webrtc::RtpTransceiverInit transceiver_init;
    transceiver_init.direction = init.direction;

    for (auto stream_id : init.stream_ids)
    {
        transceiver_init.stream_ids.push_back(std::string(stream_id.c_str()));
    }

    RTC_LOG(LS_VERBOSE) << "ArcasPeerConnection::add_video_transceiver_with_track "
                        << transceiver_init.stream_ids.size();
    auto result = api->AddTransceiver(track->ref(), transceiver_init);
    if (result.ok())
    {
        return std::make_unique<ArcasRTPVideoTransceiver>(*api, result.MoveValue());
    }

    // TODO: Handle error cases.
    return nullptr;
}

std::unique_ptr<ArcasRTPAudioTransceiver>
ArcasPeerConnection::add_audio_transceiver_with_track(std::unique_ptr<ArcasAudioTrack> track,
                                                      ArcasTransceiverInit init) const
{
    webrtc::RtpTransceiverInit transceiver_init;
    transceiver_init.direction = init.direction;

    for (auto stream_id : init.stream_ids)
    {
        transceiver_init.stream_ids.push_back(std::string(stream_id.c_str()));
    }

    RTC_LOG(LS_VERBOSE) << "ArcasPeerConnection::add_video_transceiver_with_track "
                        << transceiver_init.stream_ids.size();
    auto result = api->AddTransceiver(track->ref(), transceiver_init);
    if (result.ok())
    {
        return std::make_unique<ArcasRTPAudioTransceiver>(*api, result.MoveValue());
    }

    // TODO: Handle error cases.
    return nullptr;
}


std::unique_ptr<ArcasRTPAudioTransceiver> ArcasPeerConnection::add_audio_transceiver() const
{
    auto result = api->AddTransceiver(cricket::MEDIA_TYPE_AUDIO);

    if (result.ok())
    {
        return std::make_unique<ArcasRTPAudioTransceiver>(*api, result.MoveValue());
    }

    // TODO: Handle error cases.
    return nullptr;
}

void ArcasPeerConnection::get_stats(rust::Box<ArcasRustRTCStatsCollectorCallback> cb) const
{
    auto cb_ = rtc::make_ref_counted<ArcasRTCStatsCollectorCallback>(std::move(cb));
    api->GetStats(cb_);
}

void ArcasPeerConnection::add_ice_candidate(std::unique_ptr<ArcasICECandidate> candidate) const
{
    std::string sdp_out;
    webrtc::SdpParseError error;
    // XXX: Do *not* make further calls into candidate after this.
    auto jsep = candidate->take();
    jsep->ToString(&sdp_out);
    auto add_candidate = std::unique_ptr<webrtc::IceCandidateInterface>(
        webrtc::CreateIceCandidate(jsep->sdp_mid(), jsep->sdp_mline_index(), sdp_out, &error));

    api->AddIceCandidate(std::move(add_candidate),
                         [](webrtc::RTCError err)
                         {
                             if (!err.ok())
                             {
                                 RTC_LOG(LS_ERROR)
                                     << "ArcasPeerConnection::add_ice_candidate error: "
                                     << err.message();
                             }
                         });
}

std::unique_ptr<std::vector<ArcasRTPTransceiver>> ArcasPeerConnection::get_transceivers() const
{
    auto transceivers = api->GetTransceivers();
    std::vector<ArcasRTPTransceiver> result;
    for (auto txrx : transceivers) { result.emplace_back(*api, txrx); }
    return std::make_unique<std::vector<ArcasRTPTransceiver>>(std::move(result));
}

std::unique_ptr<ArcasDataChannel>
ArcasPeerConnection::create_data_channel(rust::String label,
                                         const ArcasDataChannelInit& init_rust) const
{
    webrtc::DataChannelInit init;
    const std::string& cxx_label = label.c_str();

    init.reliable = init_rust.reliable;
    init.ordered = init_rust.ordered;

    if (init_rust.max_retransmit_time.size() > 0)
    {
        init.maxRetransmitTime = init_rust.max_retransmit_time[0];
    }

    if (init_rust.max_retransmits.size() > 0)
    {
        init.maxRetransmits = init_rust.max_retransmits[0];
    }

    init.protocol.assign(init_rust.protocol.data(), init_rust.protocol.size());

    if (init_rust.id.size() > 0)
    {
        init.id = init_rust.id[0];
    }

    if (init_rust.priority.size() > 0)
    {
        init.priority = init_rust.priority[0];
    }

    return std::make_unique<ArcasDataChannel>(api->CreateDataChannel(cxx_label, &init));
}

File: libwebrtc-sys/src/peer_connection.rs
use cxx::UniquePtr;

use self::ffi::{ArcasRTCError, ArcasSessionDescription};

#[cxx::bridge]
pub mod ffi {

    struct ArcasTransceiverInit {
        stream_ids: Vec<String>,
        direction: ArcasCxxRtpTransceiverDirection,
    }

    // stats types
    #[derive(Debug)]
    struct ArcasVideoReceiverStats {
        pub ssrc: u32,
        pub packets_received: u32,
        pub packets_lost: i32,
        pub packets_repaired: u32,
        pub bytes_received: u64,
        pub frames_decoded: u32,
        pub keyframes_decoded: u32,
        pub frames_dropped: u32,
        pub total_decode_time: f64,
        pub frame_width: u32,
        pub frame_height: u32,
    }

    #[derive(Debug)]
    struct ArcasVideoSenderStats {
        pub ssrc: u32,
        pub packets_sent: u32,
        pub bytes_sent: u64,
        pub frames_encoded: u32,
        pub key_frames_encoded: u32,
        pub total_encode_time: f64,
        pub frame_width: u32,
        pub frame_height: u32,
        pub retransmitted_packets_sent: u64,
        pub retransmitted_bytes_sent: u64,
        pub total_packet_send_delay: f64,
        pub nack_count: u32,
        pub fir_count: u32,
        pub pli_count: u32,
        pub quality_limitation_reason: u32, // 0 - kNone, 1 - kCpu, 2 - kBandwidth, 3 - kOther
        pub quality_limitation_resolution_changes: u32,
        pub remote_packets_lost: i32,
        pub remote_jitter: f64,
        pub remote_round_trip_time: f64,
    }

    #[derive(Debug)]
    struct ArcasAudioReceiverStats {
        pub ssrc: u32,
        pub packets_received: u32,
        pub packets_lost: i32,
        pub bytes_received: u64,
        pub total_samples_received: u64,
        pub total_samples_duration: f64,
        pub audio_level: f64,
        pub total_audio_energy: f64,
    }

    #[derive(Debug)]
    struct ArcasAudioSenderStats {
        pub ssrc: u32,
        pub packets_sent: u32,
        pub bytes_sent: u64,
        pub remote_packets_lost: i32,
        pub remote_jitter: f64,
        pub remote_round_trip_time: f64,
        pub audio_level: f64,
        pub total_audio_energy: f64,
    }

    unsafe extern "C++" {
        include!("include/alias.h");
        include!("include/peer_connection_stats_callback.h");
        include!("include/peer_connection.h");
        include!("libwebrtc-sys/src/video_track.rs.h");
        type ArcasPeerConnection;
        type ArcasSDPSemantics = crate::shared_bridge::ffi::ArcasSDPSemantics;
        type ArcasRTCConfiguration = crate::shared_bridge::ffi::ArcasRTCConfiguration;
        type ArcasSessionDescription = crate::session_description::ffi::ArcasSessionDescription;
        type ArcasRTPVideoTransceiver = crate::rtp_transceiver::ffi::ArcasRTPVideoTransceiver;
        type ArcasRTPAudioTransceiver = crate::rtp_transceiver::ffi::ArcasRTPAudioTransceiver;
        type ArcasRTPTransceiver = crate::rtp_transceiver::ffi::ArcasRTPTransceiver;
        type ArcasVideoTrack = crate::video_track::ffi::ArcasVideoTrack;
        type ArcasICECandidate = crate::ice_candidate::ffi::ArcasICECandidate;
        type ArcasRTCError = crate::error::ffi::ArcasRTCError;
        type ArcasCxxRtpTransceiverDirection =
            crate::shared_bridge::ffi::ArcasCxxRtpTransceiverDirection;
        type ArcasRTCStatsCollectorCallback;
        type ArcasAudioTrack = crate::audio_track::ffi::ArcasAudioTrack;

        fn gen_shared_peer_connection() -> SharedPtr<ArcasPeerConnection>;

        // ArcasPeerConnection
        fn create_offer(
            self: &ArcasPeerConnection,
            observer: Box<ArcasRustCreateSessionDescriptionObserver>,
        );
        fn create_answer(
            self: &ArcasPeerConnection,
            observer: Box<ArcasRustCreateSessionDescriptionObserver>,
        );
        fn set_local_description(
            self: &ArcasPeerConnection,
            observer: Box<ArcasRustSetSessionDescriptionObserver>,
            session: UniquePtr<ArcasSessionDescription>,
        );

        fn set_remote_description(
            self: &ArcasPeerConnection,
            observer: Box<ArcasRustSetSessionDescriptionObserver>,
            session: UniquePtr<ArcasSessionDescription>,
        );

        fn add_video_transceiver(self: &ArcasPeerConnection)
            -> UniquePtr<ArcasRTPVideoTransceiver>;

        fn add_video_transceiver_with_track(
            self: &ArcasPeerConnection,
            track: UniquePtr<ArcasVideoTrack>,
            init: ArcasTransceiverInit,
        ) -> UniquePtr<ArcasRTPVideoTransceiver>;

        fn add_audio_transceiver_with_track(
            self: &ArcasPeerConnection,
            track: UniquePtr<ArcasAudioTrack>,
            init: ArcasTransceiverInit,
        ) -> UniquePtr<ArcasRTPAudioTransceiver>;

        fn add_audio_transceiver(self: &ArcasPeerConnection)
            -> UniquePtr<ArcasRTPAudioTransceiver>;

        fn add_video_track(
            self: &ArcasPeerConnection,
            track: UniquePtr<ArcasVideoTrack>,
            stream_ids: Vec<String>,
        );

        fn add_audio_track(
            self: &ArcasPeerConnection,
            track: UniquePtr<ArcasAudioTrack>,
            stream_ids: Vec<String>,
        );

        fn get_stats(self: &ArcasPeerConnection, callback: Box<ArcasRustRTCStatsCollectorCallback>);
        fn add_ice_candidate(self: &ArcasPeerConnection, candidate: UniquePtr<ArcasICECandidate>);
        fn close(self: &ArcasPeerConnection);
        fn get_transceivers(
            self: &ArcasPeerConnection,
        ) -> UniquePtr<CxxVector<ArcasRTPTransceiver>>;

    }

    extern "Rust" {
        type ArcasRustCreateSessionDescriptionObserver;
        type ArcasRustSetSessionDescriptionObserver;
        type ArcasRustRTCStatsCollectorCallback;
        fn on_stats_delivered(
            self: &ArcasRustRTCStatsCollectorCallback,
            video_rx: Vec<ArcasVideoReceiverStats>,
            audio_rx: Vec<ArcasAudioReceiverStats>,
            video_tx: Vec<ArcasVideoSenderStats>,
            audio_tx: Vec<ArcasAudioSenderStats>,
        );

        // ArcasRustCreateSessionDescriptionObserver
        fn on_success(
            self: &ArcasRustCreateSessionDescriptionObserver,
            success: UniquePtr<ArcasSessionDescription>,
        );
        fn on_failure(
            self: &ArcasRustCreateSessionDescriptionObserver,
            failure: UniquePtr<ArcasRTCError>,
        );
        // ArcasRustSetSessionDescriptionObserver
        fn on_success(self: &ArcasRustSetSessionDescriptionObserver);
        fn on_failure(
            self: &ArcasRustSetSessionDescriptionObserver,
            failure: UniquePtr<ArcasRTCError>,
        );
    }
}

pub struct ArcasRustCreateSessionDescriptionObserver {
    success: Box<dyn Fn(UniquePtr<ArcasSessionDescription>)>,
    failure: Box<dyn Fn(UniquePtr<ArcasRTCError>)>,
}

impl ArcasRustCreateSessionDescriptionObserver {
    pub fn new(
        success: Box<dyn Fn(UniquePtr<ArcasSessionDescription>)>,
        failure: Box<dyn Fn(UniquePtr<ArcasRTCError>)>,
    ) -> Self {
        Self { success, failure }
    }

    fn on_success(&self, description: UniquePtr<ArcasSessionDescription>) {
        (self.success)(description);
    }
    fn on_failure(&self, err: UniquePtr<ArcasRTCError>) {
        (self.failure)(err);
    }
}

pub struct ArcasRustSetSessionDescriptionObserver {
    success: Box<dyn Fn()>,
    failure: Box<dyn Fn(UniquePtr<ArcasRTCError>)>,
}

impl ArcasRustSetSessionDescriptionObserver {
    pub fn new(
        success: Box<dyn Fn()>,
        failure: Box<dyn Fn(UniquePtr<self::ffi::ArcasRTCError>)>,
    ) -> Self {
        Self { success, failure }
    }

    fn on_success(&self) {
        (self.success)();
    }
    fn on_failure(&self, err: UniquePtr<self::ffi::ArcasRTCError>) {
        (self.failure)(err);
    }
}

type StatsCallbackFn = dyn Fn(
    Vec<self::ffi::ArcasVideoReceiverStats>,
    Vec<self::ffi::ArcasAudioReceiverStats>,
    Vec<self::ffi::ArcasVideoSenderStats>,
    Vec<self::ffi::ArcasAudioSenderStats>,
);

pub struct ArcasRustRTCStatsCollectorCallback {
    cb: Box<StatsCallbackFn>,
}

impl ArcasRustRTCStatsCollectorCallback {
    pub fn new(cb: Box<StatsCallbackFn>) -> Self {
        Self { cb }
    }

    fn on_stats_delivered(
        self: &ArcasRustRTCStatsCollectorCallback,
        video_rx: Vec<self::ffi::ArcasVideoReceiverStats>,
        audio_rx: Vec<self::ffi::ArcasAudioReceiverStats>,
        video_tx: Vec<self::ffi::ArcasVideoSenderStats>,
        audio_tx: Vec<self::ffi::ArcasAudioSenderStats>,
    ) {
        (self.cb)(video_rx, audio_rx, video_tx, audio_tx)
    }
}
